networking-calico/networking_calico/agent/dhcp_agent.py

```diff
@@ -148,6 +154,146 @@ def split_endpoint_name(name):
     return tuple([p.replace('#', '-') for p in parts])
 
 
+class MTUWatcher(object):
+
+    # -----------------------------------------------------------------
+    # Methods called from DHCP agent's main thread.
+    # -----------------------------------------------------------------
+
+    def __init__(self, agent, port_handler):
+        self.agent = agent
+        self.port_handler = port_handler
+        self.mtu_by_if_name = {}
+        self.port_id_by_if_name = {}
+        self.rx_up = re.compile('[0-9]+: ([a-z0-9-]+).* mtu ([0-9]+) .* state UP')
+        self.rx_del = re.compile('Deleted [0-9]+: ([a-z0-9-]+)')
+
+    def get_mtu(self, if_name):
+        # Note, returns None if we haven't yet seen an MTU for the given
+        # interface name.
+        return self.mtu_by_if_name.get(if_name)
+
+    def watch_port(self, id, if_name):
+        self.port_id_by_if_name[if_name] = id
+
+    def unwatch_port(self, id, if_name):
+        if if_name in self.port_id_by_if_name:
+            del self.port_id_by_if_name[if_name]
+
+    def start_up(self):
+        # Start 'ip monitor link' running on a separate thread.
+        self.ip_monitor_running = False
+        eventlet.spawn(self.process_command, ['ip', 'monitor', 'link'])
```

Comment:
Perhaps `pyroute2` may handle cases like this somewhat more robustly? (caveat: I don't have practical experience with the link-monitoring aspects of it in particular).

`pyroute2` can also put the processing in another thread for you and reduce some of the boilerplate:

```python
from pyroute2 import IPRoute

with IPRoute() as ipr:
  ipr.bind(async_cache=True)

  while True:
    print('looping...')
    for message in ipr.get():
      print(message)
```

When booting a VM on a hypervisor while running this I see, e.g.:
```
{'family': 0, '__align': (), 'ifi_type': 1, 'index': 40, 'flags': 69699, 'change': 0, 'attrs': [('IFLA_IFNAME', 'tap102c5c63-3e'), ('IFLA_TXQLEN', 1000), ('IFLA_OPERSTATE', 'UP'), ('IFLA_LINKMODE', 0), ('IFLA_MTU', 1500), ... 'state': 'up', 'event': 'RTM_NEWLINK'}
```

Reply:
To your point, though, even the `pyroute2` docs mention this concern:
https://github.com/svinota/pyroute2/blob/bc1a0ea1c5cfeebd9337afecb435d4296663295d/pyroute2/netlink/nlsocket.py#L65

The intended use-case (creating/deleting taps for a hypervisor) should be such that the risk of overloading the socket buffer is OK, though, I would think...

Reply:
Good catch, I have refactored a little to handle this.

Reply:
Whether we use ip link or pyroute2 I think we need to handle ENOBUF.  You _might_ never see it with only monitoring link events but _might_ is not good enough. It can be caused by general overload of the system in turn starving our process and we don't have any control over the system.